File: 160_IntersectionOfTwoLinkedLists.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
//12.86%
struct ListNode *getIntersectionNode(struct ListNode *headA, struct ListNode *headB) {
    if(!headA || !headB)
		return NULL;
	int lenA = 0, lenB = 0;
	struct ListNode *ptrA = headA, *ptrB = headB;
	while(ptrA){
		lenA++;
		ptrA = ptrA -> next;
	}
	while(ptrB){
		lenB++;
		ptrB = ptrB -> next;
	}
	//Ensure headA points to the longer list
	int diffLen = 0;
	if(lenA > lenB){
		diffLen = lenA - lenB;
	}else{
		diffLen = lenB - lenA;
		struct ListNode *ptrTemp = headA;
		headA = headB;
		headB = ptrTemp;
	}
	
	//Find the common node out
	ptrA = headA, ptrB = headB;
	while(diffLen--)
		ptrA = ptrA -> next;
	while(ptrA && ptrB){
		if(ptrA == ptrB)
			return ptrA;
		ptrA = ptrA -> next;
		ptrB = ptrB -> next;
	}
	return NULL;
}

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
// My old solution before which clearly better
struct ListNode *getIntersectionNode(struct ListNode *headA, struct ListNode *headB) 
{
    if(NULL == headA || NULL == headB)
		return NULL;
	
	//Two pointers ptrA & ptrB
	struct ListNode *ptrA, *ptrB;
	ptrA = headA;
	ptrB = headB;
	int flagA = 0;
	int flagB = 0;
	while(ptrA != ptrB)
	{
		if(NULL == ptrA -> next)
		{
			if(0 == flagA)
			{
				ptrA = headB;
				flagA = 1;
			}
			else
				return NULL;
		}
		else
			ptrA = ptrA -> next;

		if(NULL == ptrB -> next)
		{
			if(0 == flagB)
			{
				ptrB = headA;
				flagB = 1;
			}
			else
				return NULL;
		}
		else
			ptrB = ptrB -> next;
	}
	
	return ptrA;
}

File: 160_IntersectionOfTwoLinkedLists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None
# 57.1%
class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        if not headA or not headB:
            return None
        currA, currB = headA, headB
        flagA, flagB = 0, 0
        while currA is not currB:
            if not currA.next:
                if not flagA:
                    flagA = 1
                    currA = headB
                else:
                    return None
            else:
                currA = currA.next
                
            if not currB.next:
                if not flagB:
                    flagB = 1
                    currB = headA
                else:
                    return None
            else:
                currB = currB.next
        return currA
		

# better solution from LeetCode Discuss		
 # @param two ListNodes
    # @return the intersected ListNode
    def getIntersectionNode(self, headA, headB):
        if headA is None or headB is None:
            return None

        pa = headA # 2 pointers
        pb = headB

        while pa is not pb:
            # if either pointer hits the end, switch head and continue the second traversal, 
            # if not hit the end, just move on to next
            pa = headB if pa is None else pa.next
            pb = headA if pb is None else pb.next

        return pa # only 2 ways to get out of the loop, they meet or the both hit the end=None